File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        int[] dy = {1,0,-1,0};
        int[] dx = {0,1,0,-1};
        int n = maps.length;
        int m = maps[0].length;
        int[][] visited = new int[n][m];
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{0,0});
        visited[0][0]= 1;
        while(!q.isEmpty()){
            int[] yx = q.poll();
            int y = yx[0];
            int x = yx[1];
            
            for (int i=0; i<4; i++){
                int ny = y + dy[i];
                int nx = x + dx[i];
                
                if (ny < 0 || nx < 0 || ny >= n || nx >= m || visited[ny][nx] != 0 || maps[ny][nx] == 0)continue;
                visited[ny][nx] = visited[y][x] + 1;
                q.add(new int[]{ny,nx});
            }
        }
        answer = visited[n-1][m-1];
        if(answer == 0) answer = -1;
        
        return answer;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>

using namespace std;

int dy[4] = {1,0,-1,0};
int dx[4] = {0,1,0,-1};
int visited[102][102];

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    int n = maps.size();
    int m = maps[0].size();
    queue<pair<int,int>> q;
    q.push({0,0});
    visited[0][0] = 1;
    while(q.size()){
        int y = q.front().first;
        int x = q.front().second;
        
        q.pop();
        
        for (int i=0; i<4; i++){
            int ny = y + dy[i];
            int nx = x + dx[i];
            
            if(ny < 0 || nx < 0 || ny >= n || nx >= m || maps[ny][nx] == 0 || visited[ny][nx])continue;
            visited[ny][nx] = visited[y][x] + 1;
            q.push({ny,nx});
        }
    }
    if(visited[n-1][m-1])answer = visited[n-1][m-1];
    else answer = -1;
    return answer;
}